knative-operator/pkg/controller/knativekafka/knativekafka_controller_test.go

```diff
@@ -0,0 +1,500 @@
+				{Name: "kafka-controller-manager", Namespace: "knative-sources"},
+			},
+		},
+		{
+			name:     "Create CR with channel disabled and source enabled",
+			instance: makeCr(false, true, false),
+			exists: []types.NamespacedName{
+				{Name: "kafka-controller-manager", Namespace: "knative-sources"},
+			},
+			doesNotExist: []types.NamespacedName{
+				{Name: "kafka-ch-controller", Namespace: "knative-eventing"},
+			},
+		},
+		{
+			name:     "Create CR with channel and source disabled",
+			instance: makeCr(false, false, false),
+			exists:   []types.NamespacedName{},
+			doesNotExist: []types.NamespacedName{
+				{Name: "kafka-ch-controller", Namespace: "knative-eventing"},
+				{Name: "kafka-controller-manager", Namespace: "knative-sources"},
+			},
+		},
+		{
+			name:     "Delete CR",
+			instance: makeCr(true, true, true),
+			exists:   []types.NamespacedName{},
+			doesNotExist: []types.NamespacedName{
+				{Name: "kafka-ch-controller", Namespace: "knative-eventing"},
+				{Name: "kafka-controller-manager", Namespace: "knative-sources"},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			// Register operator types with the runtime scheme.
+			s := scheme.Scheme
+			s.AddKnownTypes(v1alpha1.SchemeGroupVersion, test.instance)
+
+			initObjs := []runtime.Object{test.instance}
+
+			cl := fake.NewFakeClient(initObjs...)
+			mgr := &fakeManager{
+				client: cl,
+				scheme: s,
+			}
+			r, err := newReconciler(mgr)
+
+			// Reconcile to intialize
+			if _, err := r.Reconcile(defaultRequest); err != nil {
+				t.Fatalf("reconcile: (%v)", err)
+			}
+
+			// check if things that should exist is created
+			for _, d := range test.exists {
+				deployment := &appsv1.Deployment{}
+				err = cl.Get(context.TODO(), d, deployment)
+				if err != nil {
+					t.Fatalf("get: (%v)", err)
+				}
+			}
+
+			// check if things that shouldnot exist is deleted
+			for _, d := range test.doesNotExist {
+				deployment := &appsv1.Deployment{}
+				err = cl.Get(context.TODO(), d, deployment)
+				if err == nil || !errors.IsNotFound(err) {
+					t.Fatalf("exists: (%v)", err)
+				}
+			}
+
+			// delete deployments to see if they're recreated
+			for _, d := range test.exists {
+				deployment := &appsv1.Deployment{}
+				err = cl.Get(context.TODO(), d, deployment)
+				if err != nil {
+					t.Fatalf("get: (%v)", err)
+				}
+				err = cl.Delete(context.TODO(), deployment)
+				if err != nil {
+					t.Fatalf("delete: (%v)", err)
+				}
+			}
+
+			// Reconcile again
+			if _, err := r.Reconcile(defaultRequest); err != nil {
+				t.Fatalf("reconcile: (%v)", err)
+			}
+
+			// check if things that should exist is created
+			for _, d := range test.exists {
+				deployment := &appsv1.Deployment{}
+				err = cl.Get(context.TODO(), d, deployment)
+				if err != nil {
+					t.Fatalf("get: (%v)", err)
+				}
+			}
+		})
+	}
+}
+
+func makeCr(channel bool, source bool, deleted bool) *v1alpha1.KnativeKafka {
```

Comment:
nit: Instead of using booleans here, function options might be a great way to make the tests a bit more expressive.

```golang
func kk(mods ...func(*v1alpha1.KnativeKafka)) *v1alpha1.KnativeKafka
  base := // create a minimal CR
  for _, mod := range mods {
    mod(base)
  }
  return base
}

func withSourceEnabled(kk *v1alpha1.KnativeKafka) {
  kk.Spec.Source.Enabled = true
}

// etc. withChannelEnabled, withDeletionTimestamp

// Usage then goes like
kk()
kk(withChannelEnabled)
kk(withDeletionTimestamp)
kk(withChannelEnabled, withSourceEnabled)

Reply:
I think your suggestion is good, but I don't like the fact that it is not obvious as the boolean arguments.
```
kk() // --> what does this mean? all disabled? all enabled? superstate?
kk(withChannelEnabled, withSourceEnabled) // --> easy to understand, both are enabled
kk(withChannelEnabled) // --> what does this mean? source is enabled or disabled?
```

Reply:
Well you have to define the default state, which is all disabled (because the structs default to disabled).

You can of course also provide the respective `withChannelDisabled` etc. for clarity.

Reply:
But I marked it as a nit, so feel free to ignore :)